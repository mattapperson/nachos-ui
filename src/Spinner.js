import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, Animated, Easing } from 'react-native'
import themeManager from './themeManager'

const defaultTheme = {
  SPINNER_SIZE: 25,
  SPINNER_COLOR: '#2f8cff',
  SPINNER_DURATION: 450,
}

class Spinner extends Component {
  static defaultStyle = (theme = defaultTheme) => {
    return {
      base: { flexDirection: 'row' },
    }
  }
  static propTypes = {
    duration: PropTypes.number,
    size: PropTypes.number,
    color: PropTypes.string,
    theme: PropTypes.object,
  }

  static defaultProps = { duration: 450 }

  constructor(props) {
    super(props)
    this.firstAnimatedValue = new Animated.Value(0)
    this.secondAnimatedValue = new Animated.Value(0)
    this.thirdAnimatedValue = new Animated.Value(0)
    this._shouldAnimationStop = false
  }

  componentDidMount() {
    this._shouldAnimationStop = false
    this._setupAnimation()
  }

  componentWillUnmount() {
    this._shouldAnimationStop = true
  }

  _setupAnimation = () => {
    const createAnimation = (value) => {
      return Animated.sequence([Animated.timing(value, {
        toValue: 1,
        duration: this.props.duration,
        easing: Easing.ease,
      }), Animated.timing(value, {
        toValue: 0,
        duration: this.props.duration,
        easing: Easing.ease,
      })])
    }

    const scheduleAnimation = (value) => {
      createAnimation(value).start()
      return new Promise((resolve) => {
        setTimeout(
          () => {
            resolve()
          },
          this.props.duration
        )
      })
    }

    const animate = () => {
      if (this._shouldAnimationStop) {
        return
      }
      scheduleAnimation(this.firstAnimatedValue)
        .then(() => scheduleAnimation(this.secondAnimatedValue))
        .then(() => scheduleAnimation(this.thirdAnimatedValue))
        .then(() => animate())
    }

    animate()
  }

  _interpolate = (value) => {
    const scale = value.interpolate({
      inputRange: [0, 1],
      outputRange: [0.4, 1],
    })
    return scale
  }

  render() {
    const theme = this.props.theme

    const spinnerItem = {
      width: this.props.size || theme.SPINNER_SIZE,
      height: this.props.size || theme.SPINNER_SIZE,
      borderRadius: this.props.size || theme.SPINNER_SIZE,
      backgroundColor: this.props.color || theme.SPINNER_COLOR,
    }
    const firstScale = this._interpolate(this.firstAnimatedValue)
    const secondScale = this._interpolate(this.secondAnimatedValue)
    const thirdScale = this._interpolate(this.thirdAnimatedValue)

    return (
      <View style={theme.base}>
        <Animated.View
          style={[
            spinnerItem,
            { transform: [{ scale: firstScale }] },
          ]}
        />
        <Animated.View
          style={[
            spinnerItem,
            { transform: [{ scale: secondScale }] },
          ]}
        />
        <Animated.View
          style={[
            spinnerItem,
            { transform: [{ scale: thirdScale }] },
          ]}
        />
      </View>
    )
  }
}

export default Spinner
