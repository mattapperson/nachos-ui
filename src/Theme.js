import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
