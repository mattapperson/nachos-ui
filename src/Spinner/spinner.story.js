import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import React from "react";
import Spinner from "./index";

import { View } from "react-native";

const Example = () => {
  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Spinner />
    </View>
  );
};

storiesOf("Spinner", module).add("default", withInfo()(Example));

export default Example;
