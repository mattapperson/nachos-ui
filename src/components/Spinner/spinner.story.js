import { storiesOf } from "../../storybook";
import { withInfo } from "../../storybook";
import React from "react";
import Spinner from "./index";

import { View } from "react-native";

const Example = () => {
  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Spinner />
    </View>
  );
};

storiesOf("Spinner", module).add("default", withInfo()(Example));

export default Example;
