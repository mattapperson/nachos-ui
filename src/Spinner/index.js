import React, { Component } from "react";
import PropTypes from "prop-types";
import { View, Animated, Easing } from "react-native";
import { withTheme } from "../Theme";

class Spinner extends Component {
  static themeConfig = {
    props: {
      size: 25,
      color: "@primaryColor",
      duration: 450
    },
    style: {
      base: { flexDirection: "row" }
    }
  };
  static propTypes = {
    duration: PropTypes.number,
    size: PropTypes.number,
    color: PropTypes.string,
    theme: PropTypes.object
  };

  static defaultProps = { duration: 450 };

  constructor(props) {
    super(props);
    this.firstAnimatedValue = new Animated.Value(0);
    this.secondAnimatedValue = new Animated.Value(0);
    this.thirdAnimatedValue = new Animated.Value(0);
    this._shouldAnimationStop = false;
  }

  componentDidMount() {
    this._shouldAnimationStop = false;
    this._setupAnimation();
  }

  componentWillUnmount() {
    this._shouldAnimationStop = true;
  }

  _setupAnimation = () => {
    const createAnimation = value => {
      return Animated.sequence([
        Animated.timing(value, {
          toValue: 1,
          duration: this.props.duration,
          easing: Easing.ease
        }),
        Animated.timing(value, {
          toValue: 0,
          duration: this.props.duration,
          easing: Easing.ease
        })
      ]);
    };

    const scheduleAnimation = value => {
      createAnimation(value).start();
      return new Promise(resolve => {
        setTimeout(() => {
          resolve();
        }, this.props.duration);
      });
    };

    const animate = () => {
      if (this._shouldAnimationStop) {
        return;
      }
      scheduleAnimation(this.firstAnimatedValue)
        .then(() => scheduleAnimation(this.secondAnimatedValue))
        .then(() => scheduleAnimation(this.thirdAnimatedValue))
        .then(() => animate());
    };

    animate();
  };

  _interpolate = value => {
    const scale = value.interpolate({
      inputRange: [0, 1],
      outputRange: [0.4, 1]
    });
    return scale;
  };

  render() {
    const theme = this.props.theme;

    const spinnerItem = {
      width: this.props.size,
      height: this.props.size,
      borderRadius: this.props.size,
      backgroundColor: this.props.color
    };
    const firstScale = this._interpolate(this.firstAnimatedValue);
    const secondScale = this._interpolate(this.secondAnimatedValue);
    const thirdScale = this._interpolate(this.thirdAnimatedValue);

    return (
      <View style={theme.base}>
        <Animated.View
          style={[spinnerItem, { transform: [{ scale: firstScale }] }]}
        />
        <Animated.View
          style={[spinnerItem, { transform: [{ scale: secondScale }] }]}
        />
        <Animated.View
          style={[spinnerItem, { transform: [{ scale: thirdScale }] }]}
        />
      </View>
    );
  }
}

export default withTheme("Spinner", Spinner);
