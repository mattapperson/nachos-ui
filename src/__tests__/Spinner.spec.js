import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import Spinner from "../Spinner";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Spinner />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
