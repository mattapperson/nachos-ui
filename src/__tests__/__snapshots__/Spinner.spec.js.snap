// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
>
  <div
    className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
    style={
      Object {
        "WebkitTransform": "scale(0.4)",
        "backgroundColor": "rgba(47,140,255,1)",
        "borderBottomLeftRadius": "25px",
        "borderBottomRightRadius": "25px",
        "borderTopLeftRadius": "25px",
        "borderTopRightRadius": "25px",
        "height": "25px",
        "transform": "scale(0.4)",
        "width": "25px",
      }
    }
  />
  <div
    className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
    style={
      Object {
        "WebkitTransform": "scale(0.4)",
        "backgroundColor": "rgba(47,140,255,1)",
        "borderBottomLeftRadius": "25px",
        "borderBottomRightRadius": "25px",
        "borderTopLeftRadius": "25px",
        "borderTopRightRadius": "25px",
        "height": "25px",
        "transform": "scale(0.4)",
        "width": "25px",
      }
    }
  />
  <div
    className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
    style={
      Object {
        "WebkitTransform": "scale(0.4)",
        "backgroundColor": "rgba(47,140,255,1)",
        "borderBottomLeftRadius": "25px",
        "borderBottomRightRadius": "25px",
        "borderTopLeftRadius": "25px",
        "borderTopRightRadius": "25px",
        "height": "25px",
        "transform": "scale(0.4)",
        "width": "25px",
      }
    }
  />
</div>
`;
